File: src/main/java/io/polygloat/controllers/TranslationController.java
package io.polygloat.controllers;

import io.polygloat.dtos.PathDTO;
import io.polygloat.dtos.request.SetTranslationsDTO;
import io.polygloat.dtos.response.KeyResponseDTO;
import io.polygloat.dtos.response.ViewDataResponse;
import io.polygloat.dtos.response.translations_view.ResponseParams;
import io.polygloat.exceptions.NotFoundException;
import io.polygloat.model.Key;
import io.polygloat.model.Permission;
import io.polygloat.service.KeyService;
import io.polygloat.service.RepositoryService;
import io.polygloat.service.SecurityService;
import io.polygloat.service.TranslationService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

@RestController
@CrossOrigin(origins = "*")
@RequestMapping("/api/repository/{repositoryId}/translations")
public class TranslationController implements IController {

    private final TranslationService translationService;
    private final KeyService keyService;
    private final RepositoryService repositoryService;
    private final SecurityService securityService;

    @Autowired
    public TranslationController(TranslationService translationService, KeyService keyService, RepositoryService repositoryService, SecurityService securityService) {
        this.translationService = translationService;
        this.keyService = keyService;
        this.repositoryService = repositoryService;
        this.securityService = securityService;
    }


    @GetMapping(value = "/{languages}")
    public Map<String, Object> getTranslations(@PathVariable("repositoryId") Long repositoryId,
                                               @PathVariable("languages") Set<String> languages) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.VIEW);
        return translationService.getTranslations(languages, repositoryId);
    }

    @PostMapping("/set")
    public void setTranslations(@PathVariable("repositoryId") Long repositoryId, @RequestBody @Valid SetTranslationsDTO dto) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.TRANSLATE);
        Key key = keyService.get(repositoryId, PathDTO.fromFullPath(dto.getKey())).orElseThrow(NotFoundException::new);
        translationService.setForKey(key, dto.getTranslations());
    }

    @GetMapping(value = "/view")
    public ViewDataResponse<LinkedHashSet<KeyResponseDTO>, ResponseParams> getViewData(@PathVariable("repositoryId") Long repositoryId,
                                                                                       @RequestParam(name = "languages", required = false) Set<String> languages,
                                                                                       @RequestParam(name = "limit", defaultValue = "10") int limit,
                                                                                       @RequestParam(name = "offset", defaultValue = "0") int offset,
                                                                                       @RequestParam(name = "search", required = false) String search
    ) {
        securityService.checkRepositoryPermission(repositoryId, Permission.RepositoryPermissionType.VIEW);
        return translationService.getViewData(languages, repositoryId, limit, offset, search);
    }
}

File: src/main/java/io/tolgee/controllers/TranslationController.kt
@file:Suppress("MVCPathVariableInspection")

package io.tolgee.controllers

import io.swagger.v3.oas.annotations.Hidden
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import io.tolgee.constants.ApiScope
import io.tolgee.dtos.PathDTO
import io.tolgee.dtos.request.SetTranslationsDTO
import io.tolgee.dtos.response.KeyWithTranslationsResponseDto
import io.tolgee.dtos.response.ViewDataResponse
import io.tolgee.dtos.response.translations_view.ResponseParams
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Permission
import io.tolgee.security.api_key_auth.AccessWithApiKey
import io.tolgee.security.repository_auth.AccessWithAnyRepositoryPermission
import io.tolgee.security.repository_auth.AccessWithRepositoryPermission
import io.tolgee.security.repository_auth.RepositoryHolder
import io.tolgee.service.KeyService
import io.tolgee.service.RepositoryService
import io.tolgee.service.SecurityService
import io.tolgee.service.TranslationService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping("/api/repository/{repositoryId:[0-9]+}/translations", "/api/repository/translations")
@Tag(name = "Translations", description = "Manipulates localization messages and metadata")
class TranslationController @Autowired constructor(
        private val translationService: TranslationService,
        private val keyService: KeyService,
        private val securityService: SecurityService,
        private val repositoryHolder: RepositoryHolder,
        private val repositoryService: RepositoryService
) : IController {
    @GetMapping(value = ["/{languages}"])
    @AccessWithAnyRepositoryPermission
    @AccessWithApiKey(scopes = [ApiScope.TRANSLATIONS_VIEW])
    @Operation(summary = "Get all translations for specific languages")
    fun getTranslations(@PathVariable("languages") languages: Set<String?>?): Map<String, Any> {
        return translationService.getTranslations(languages, repositoryHolder.repository.id)
    }

    @Suppress("DeprecatedCallableAddReplaceWith")
    @PostMapping("/set")
    @AccessWithApiKey(scopes = [ApiScope.TRANSLATIONS_EDIT])
    @AccessWithRepositoryPermission(permission = Permission.RepositoryPermissionType.TRANSLATE)
    @Deprecated(message = "Use put method to /api/repository/{repositoryId}/translations or /api/repository/translations")
    @Hidden
    fun setTranslationsPost(@RequestBody @Valid dto: SetTranslationsDTO?) {
        setTranslations(dto)
    }

    @PutMapping("")
    @AccessWithApiKey(scopes = [ApiScope.TRANSLATIONS_EDIT])
    @AccessWithRepositoryPermission(permission = Permission.RepositoryPermissionType.TRANSLATE)
    @Operation(summary = "Sets translations for existing key")
    fun setTranslations(@RequestBody @Valid dto: SetTranslationsDTO?) {
        val key = keyService.get(
                repositoryHolder.repository.id,
                PathDTO.fromFullPath(dto!!.key)
        ).orElseThrow { NotFoundException() }

        translationService.setForKey(key, dto.translations!!)
    }

    @PostMapping("")
    @AccessWithApiKey([ApiScope.KEYS_EDIT, ApiScope.TRANSLATIONS_EDIT])
    @AccessWithRepositoryPermission(permission = Permission.RepositoryPermissionType.EDIT)
    @Operation(summary = "Sets translations for existing or not existing key")
    fun createOrUpdateTranslations(@RequestBody @Valid dto: SetTranslationsDTO) {
        val repository = repositoryService.get(repositoryHolder.repository.id).get()
        val key = keyService.getOrCreateKey(repository, PathDTO.fromFullPath(dto.key))
        translationService.setForKey(key, dto.translations!!)
    }

    @GetMapping(value = ["/view"])
    @Operation(summary = "Returns data for translations view with metadata")
    fun getViewData(@PathVariable("repositoryId") repositoryId: Long?,
                    @RequestParam(name = "languages", required = false) languages: Set<String>?,
                    @RequestParam(name = "limit", defaultValue = "10") limit: Int,
                    @RequestParam(name = "offset", defaultValue = "0") offset: Int,
                    @RequestParam(name = "search", required = false) search: String?
    ): ViewDataResponse<LinkedHashSet<KeyWithTranslationsResponseDto>, ResponseParams> {
        securityService.checkRepositoryPermission(repositoryId!!, Permission.RepositoryPermissionType.VIEW)
        return translationService.getViewData(languages, repositoryId, limit, offset, search)
    }
}
